android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java

```diff
@@ -205,7 +205,63 @@ public void setIsChiral(boolean isChiral) {
   }
 
   @JavascriptInterface
-  public abstract boolean dispatchKey(final int code, final int eventModifiers);
+  public boolean dispatchKey(final int code, final int eventModifiers) {
+    Handler mainLoop = new Handler(Looper.getMainLooper());
+    mainLoop.post(new Runnable() {
+      public void run() {
+        if (k == null) {
+          KMLog.LogError(TAG, "dispatchKey failed: Keyboard is null");
+          return;
+        }
+
+        if (k.subKeysWindow != null) {
+          return;
+        }
+
+        if (k.keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP &&
+            (KMTextView.activeView == null || KMTextView.activeView.getClass() != KMTextView.class)) {
+          if (KMTextView.activeView == null && KMManager.isDebugMode()) {
+            Log.w(TAG, "dispatchKey failed: activeView is null");
+          }
+          return;
+        }
+
+        InputConnection ic = (k.keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP) ?
+          KMTextView.activeView.onCreateInputConnection(new EditorInfo()) :
+          KMManager.getInputMethodService().getCurrentInputConnection();
```

Comment:
This feels like it should be a helper function that will probably be used elsewhere too. FWIW, I see 8 calls to `getCurrentInputConnection()` in our current master source.

Reply:
I can refactor some of the calls. The rest are used for system keyboard apps:
* Keyman
* KMSample2
* FV Android